core/tests/performance_framework/process_data.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2016 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Script to process page loading stats or HTTP Archive(also referred to as HAR)
+data and/or timing stats to obtain stats/metrics for the corresponding page.
+
+We need selenium + browsermob-proxy to capture page loading session
+(or HTTP Archive), to obtain page size stats. However, owing to the sluggish
+nature of the proxy server, timing stats are not accurate.
+Hence, we have separate timing stats which are obtained without making
+use of the proxy, directly from the browser console.
+"""
+
+
+class PageSessionMetricsRetriever(object):
+    """Contains methods to process stats and provide performance metrics.
+
+    page_session_stats if a dictionary containing page load statistics or
+    HTTP Archive (HAR).
+    Refer: https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/HAR/Overview.html
+
+    page_session_timings is a dictionary containing various timings associated
+    with a page load like request time, dom load time, etc.
+    Structure:
+        {
+          "webkitClearResourceTimings": {},
+          "memory": {...},
+          "timing": {
+            "domLoading": 1465470586738,
+            "fetchStart": 1465470583251,
+          }
+        }
+    """
+
+    def __init__(self, page_session_stats, page_session_timings):
+        self.page_session_stats = page_session_stats
+        self.page_session_timings = page_session_timings
+
+        self.page_load_timings = {}
+        if 'timing' not in self.page_session_timings:
+            raise ValueError('page_session_timings missing key: timing')
+
+        self.page_load_timings = page_session_timings['timing']
+
+    def get_request_count(self):
+        """Return no of requests made for the page load."""
+        request_count = 0
+
+        if 'log' in self.page_session_stats:
+            if 'entries' in self.page_session_stats['log']:
+                request_count = len(self.page_session_stats['log']['entries'])
+
+        if request_count == 0:
+            raise Exception('Total requests cannot be 0.')
```

Comment:
I don't think we want to test for the exceptions, though it is reasonable to raise an Exception in the situation where ``page_session_stats`` is required but not initialized. That's a different exception than here (where this one is more like a validation exception).